signer/kms/aws/signer.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2023 The Witness Contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aws
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/aws/aws-sdk-go-v2/service/kms/types"
+	"github.com/in-toto/go-witness/cryptoutil"
+	"github.com/in-toto/go-witness/log"
+	kms "github.com/in-toto/go-witness/signer/kms"
+)
+
+var awsSupportedAlgorithms = []types.CustomerMasterKeySpec{
+	types.CustomerMasterKeySpecRsa2048,
+	types.CustomerMasterKeySpecRsa3072,
+	types.CustomerMasterKeySpecRsa4096,
+	types.CustomerMasterKeySpecEccNistP256,
+	types.CustomerMasterKeySpecEccNistP384,
+	types.CustomerMasterKeySpecEccNistP521,
+}
+
+var awsSupportedHashFuncs = []crypto.Hash{
+	crypto.SHA256,
+	crypto.SHA384,
+	crypto.SHA512,
+}
+
+// SignerVerifier is a cryptoutil.SignerVerifier that uses the AWS Key Management Service
+type SignerVerifier struct {
+	reference string
+	client    client
+	hashFunc  crypto.Hash
+}
+
+// LoadSignerVerifier generates signatures using the specified key object in AWS KMS and hash algorithm.
+func LoadSignerVerifier(ctx context.Context, ksp *kms.KMSSignerProvider) (*SignerVerifier, error) {
+	a := &SignerVerifier{
+		reference: ksp.Reference,
+	}
+
+	var err error
+	a.client, err = newAWSClient(ctx, ksp)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, hashFunc := range awsSupportedHashFuncs {
+		if hashFunc == ksp.HashFunc {
+			a.hashFunc = ksp.HashFunc
+		}
+	}
+
+	if a.hashFunc == 0 {
+		return nil, fmt.Errorf("unsupported hash function: %v", ksp.HashFunc)
+	}
+
+	return a, nil
+}
+
+// NOTE: This might be all wrong but setting it like so for now
+//
+// KeyID returns the key identifier for the key used by this signer.
+func (a *SignerVerifier) KeyID() (string, error) {
+	return a.reference, nil
+}
```

Comment:
Is this the right KeyID to use? From the awsClient in client.go: https://github.com/in-toto/go-witness/pull/120/files#diff-5679de7065f19bd20365541d52c8835494a381507a05877b4f34569d62a7748fR115

Reply:
Additionally, should we have a `PublicKey()` function too?

Reply:
aha - sorry @jkjell this has gotten lost in other comments and actually poses a question:
> Should the KeyID be the KMS reference?

At the moment a lot of the functionality (e.g., defining KMS in the policy) relies on the KeyID being the KMS reference.